packages/@aws-cdk/core/test/test.staging.ts

```diff
@@ -133,10 +139,144 @@ export = {
     ]);
 
     // asset is bundled in a directory inside .cdk.staging
-    const stagingTmp = path.join('.', '.cdk.staging');
-    test.ok(ensureDirSyncSpy.calledWith(stagingTmp));
-    test.ok(mkdtempSyncSpy.calledWith(sinon.match(path.join(stagingTmp, 'asset-bundle-'))));
-    test.ok(chmodSyncSpy.calledWith(sinon.match(path.join(stagingTmp, 'asset-bundle-')), 0o777));
+    test.ok(ensureDirSyncSpy.calledWith(STAGING_TMP_DIRECTORY));
+    test.ok(ensureDirSyncSpy.calledWith(path.resolve(path.join(STAGING_TMP_DIRECTORY, 'asset-bundle-hash-dec215520dfd57a87aa1362e9e15131938583cd6e5a2bd9a45d38fe5dc5ab3d7'))));
+    test.done();
+  },
+
+  'bundler reuses its output when it can'(test: Test) {
+    // GIVEN
+    const app = new App();
+    const stack = new Stack(app, 'stack');
+    const directory = path.join(__dirname, 'fs', 'fixtures', 'test1');
+    const ensureDirSyncSpy = sinon.spy(fs, 'ensureDirSync');
+
+    // WHEN
+    new AssetStaging(stack, 'Asset', {
+      sourcePath: directory,
+      bundling: {
+        image: BundlingDockerImage.fromRegistry('alpine'),
+        command: [ DockerStubCommand.SUCCESS ],
+      },
+    });
+
+    new AssetStaging(stack, 'AssetDuplicate', {
+      sourcePath: directory,
+      bundling: {
+        image: BundlingDockerImage.fromRegistry('alpine'),
+        command: [ DockerStubCommand.SUCCESS ],
+      },
+    });
+
+    // THEN
+    app.synth();
+
+    // We're testing that docker was run exactly once even though there are two bundling assets.
+    test.deepEqual(
+      readDockerStubInputConcat(),
+      `run --rm ${USER_ARG} -v /input:/asset-input:delegated -v /output:/asset-output:delegated -w /asset-input alpine DOCKER_STUB_SUCCESS`,
+    );
+
+    // asset is bundled in a directory inside .cdk.staging
+    test.ok(ensureDirSyncSpy.calledWith(STAGING_TMP_DIRECTORY));
+    test.ok(ensureDirSyncSpy.calledWith(path.resolve(path.join(STAGING_TMP_DIRECTORY, 'asset-bundle-hash-dec215520dfd57a87aa1362e9e15131938583cd6e5a2bd9a45d38fe5dc5ab3d7'))));
+
+    test.done();
+  },
+
+  'bundler considers its options when reusing bundle output'(test: Test) {
+    // GIVEN
+    const app = new App();
+    const stack = new Stack(app, 'stack');
+    const directory = path.join(__dirname, 'fs', 'fixtures', 'test1');
+    const ensureDirSyncSpy = sinon.spy(fs, 'ensureDirSync');
+
+    // WHEN
+    new AssetStaging(stack, 'Asset', {
+      sourcePath: directory,
+      bundling: {
+        image: BundlingDockerImage.fromRegistry('alpine'),
+        command: [ DockerStubCommand.SUCCESS ],
+      },
+    });
+
+    new AssetStaging(stack, 'AssetWithDifferentBundlingOptions', {
+      sourcePath: directory,
+      bundling: {
+        image: BundlingDockerImage.fromRegistry('alpine'),
+        command: [ DockerStubCommand.SUCCESS ],
+        environment: {
+          UNIQUE_ENV_VAR: 'SOMEVALUE',
+        },
+      },
+    });
```

Comment:
Docker runs twice but the logic that we have in `synthesize()` will only move the first asset to the assembly dir (`cdk.out`):
https://github.com/aws/aws-cdk/blob/ba5b268681ccae7c7a948d93494134285e6c112b/packages/%40aws-cdk/core/lib/asset-staging.ts#L116-L126

`targetPath` is based only on the source hash in this case.

Adding an expectation that `cdk.out` should contain both assets would make the test fail here.

Reply:
Makes sense. I'll work on this now.